[dynamo, BE] Remove IncorrectUsage exception

yashwantbezawada opened the pull request:
## Summary

This PR addresses issue #167142 by removing the unused `IncorrectUsage` exception and replacing its few usages with more appropriate exception types.

## Changes

### 1. Removed `IncorrectUsage` class definition
- **File**: `torch/_dynamo/exc.py`
- **Change**: Removed the `IncorrectUsage(Exception)` class (lines 295-296)

### 2. Replaced with `RuntimeError` in `decorators.py`
- **File**: `torch/_dynamo/decorators.py`
- **Location**: `_disallow_in_graph_helper` function
- **Reasoning**: This is a user error when `disallow_in_graph` is used incorrectly (on a function that isn't already allowed)
- **Change**: `raise IncorrectUsage(...)` → `raise RuntimeError(...)`

### 3. Replaced with `unimplemented_v2` in `codegen.py`
- **File**: `torch/_dynamo/codegen.py`
- **Location**: When returning a `@contextmanager` object from a compiled function
- **Reasoning**: This is a "Not Yet Implemented" case that should trigger a graph break with helpful error messaging
- **Change**: Replaced `raise IncorrectUsage("NYI: ...")` with `unimplemented_v2(...)` providing:
  - Graph break type: `"returning_contextmanager_from_compile"`
  - Context for developers
  - User-facing explanation
  - Helpful hints

### 4. Replaced with `RuntimeError` in `higher_order_ops.py`
- **File**: `torch/_dynamo/variables/higher_order_ops.py`
- **Location**: `hints_wrapper` function
- **Reasoning**: This is an internal programming error (hints key must be provided)
- **Change**: `raise IncorrectUsage(...)` → `raise RuntimeError(...)`

### 5. Updated test
- **File**: `test/dynamo/test_decorators.py`
- **Change**: Updated `test_incorrect_usage_disallow_in_graph` to expect `RuntimeError` instead of `IncorrectUsage`

## Impact

- **Code reduction**: 3 lines removed (class definition + blank lines)
- **Better error handling**: Users now get appropriate exceptions:
  - `RuntimeError` for API misuse
  - Graph breaks with structured error messages for NYI features
- **No behavior changes**: Same errors raised in same conditions, just with more appropriate exception types

## Testing

- Syntax validated
- Existing test updated to expect `RuntimeError`
- All imports cleaned up

Fixes #167142

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo @chenyang78

yashwantbezawada committed:
[dynamo, BE] Remove IncorrectUsage exception

This change removes the unused IncorrectUsage exception and replaces its
usages with more appropriate exception types (RuntimeError or graph breaks).

Changes:
- Removed IncorrectUsage class definition from torch/_dynamo/exc.py

- torch/_dynamo/decorators.py:
  * Replaced IncorrectUsage with RuntimeError in disallow_in_graph
  * This is a user error when using the decorator incorrectly

- torch/_dynamo/codegen.py:
  * Replaced IncorrectUsage with unimplemented_v2 (graph break)
  * NYI: Returning @contextmanager from torch.compile function
  * Provides better error message with hints for users

- torch/_dynamo/variables/higher_order_ops.py:
  * Replaced IncorrectUsage with RuntimeError in hints_wrapper
  * This is an internal programming error

- test/dynamo/test_decorators.py:
  * Updated test to expect RuntimeError instead of IncorrectUsage

Fixes #167142

yashwantbezawada commented:
@pytorchbot label "topic: not user facing"

Lucaskabela reviewed (changes requested):
Generally okay but I believe `unimplemented_v2` is going under a refactor rn and should have all args passed as kwargs (cc @williamwen42 )

yashwantbezawada committed:
Update tests to expect Unsupported exception

The change in codegen.py now uses unimplemented_v2() which raises
Unsupported instead of the previous IncorrectUsage exception.

Updated three tests that return context managers from compiled functions:
- test_disable___enter__
- test_return_new_contextmanager
- test_return_advanced_contextmanager

yashwantbezawada commented:
@williamwen42 @Lucaskabela I've rebased this PR on top of the latest main (which includes PR #167255). 

Changes made:
- Updated to use the new `unimplemented()` API instead of `unimplemented_v2`
- All arguments are now passed as keyword arguments as requested
- Removed the import of `IncorrectUsage` from `codegen.py`

The code now uses:
\`\`\`python
unimplemented(
    gb_type="returning_contextmanager_from_compile",
    context="Attempting to return a @contextmanager object from a torch.compile function",
    explanation="Returning a @contextmanager object from a torch.compile function is not yet implemented",
    hints=["Consider refactoring to avoid returning context managers from compiled functions"],
)
\`\`\`

Ready for another review!

yashwantbezawada commented:
## 📝 Updated PR Description Clarification

The PR description above is outdated after the rebase. Here are the **actual changes** in this PR:

### Files Changed:
1. ✅ `torch/_dynamo/exc.py` - Removed `IncorrectUsage` class
2. ✅ `torch/_dynamo/decorators.py` - Changed `IncorrectUsage` → `RuntimeError`
3. ✅ `torch/_dynamo/codegen.py` - Changed `IncorrectUsage` → `unimplemented()` (with kwargs)
4. ✅ `test/dynamo/test_decorators.py` - Updated to expect `RuntimeError`
5. ✅ `test/dynamo/test_ctx_manager.py` - Updated 3 tests to expect `Unsupported`

### Key Corrections:
- ❌ **NOT** using `unimplemented_v2` - **NOW** using `unimplemented()` with keyword arguments
- ❌ **NOT** changing `higher_order_ops.py` - that file is not in this PR
- ✅ **NOW** includes test updates for context manager tests

### Code in `codegen.py` uses:
\`\`\`python
unimplemented(
    gb_type="returning_contextmanager_from_compile",
    context="Attempting to return a @contextmanager object from a torch.compile function",
    explanation="Returning a @contextmanager object from a torch.compile function is not yet implemented",
    hints=["Consider refactoring to avoid returning context managers from compiled functions"],
)
\`\`\`

This aligns with the refactor in #167255 as requested by @williamwen42 and @Lucaskabela.

williamwen42 commented on torch/_dynamo/codegen.py:
```diff
@@ -245,8 +245,11 @@ def __call__(
-            raise IncorrectUsage(
-                "NYI: Returning a @contextmanager object from a torch.compile function"
+            unimplemented(
+                gb_type="returning_contextmanager_from_compile",
```
```suggestion
                gb_type="returning @contextmanager object from compile",
```

williamwen42 commented on torch/_dynamo/codegen.py:
```diff
@@ -245,8 +245,11 @@ def __call__(
-                "NYI: Returning a @contextmanager object from a torch.compile function"
+            unimplemented(
+                gb_type="returning_contextmanager_from_compile",
+                context="Attempting to return a @contextmanager object from a torch.compile function",
```
```suggestion
                context=str(value)
```

williamwen42 reviewed (commented):
mainly lgtm - just a few small fixes and let's wait until CI is green

yashwantbezawada committed:
Add GB0362 entry for returning_contextmanager_from_compile graph break

yashwantbezawada committed:
Apply williamwen42's review feedback

- Update gb_type to 'returning @contextmanager object from compile'
- Change context to use str(value) for dynamic context
- Add graph_break_hints.SUPPORTABLE to hints
- Update GB0362 registry entry to match new parameters

williamwen42 reviewed (changes requested):
The changes to graph_break_registry.json are incorrect. It should just be an addition to the bottom of the file.

yashwantbezawada commented:
You're right, my apologies for the confusion. I've reverted that commit - the registry already had the correct GB0362 entry added previously. The PR should now only contain the intended changes to remove IncorrectUsage.

yashwantbezawada committed:
Fix GB0362 registry context to match actual code

The registry entry for GB0362 had Context: "{value}" but the actual
code in codegen.py line 250 uses context=str(value). This was causing
the lintrunner-noclang-partial check to fail with "Registry sync needed".

Updated the Context field to "str(value)" to match the actual call site.

Signed-off-by: Yashwant Bezawada <yashwant_b@me.com>

yashwantbezawada committed:
Fix test_disable___enter__ to expect InternalTorchDynamoError

This test calls ctx.__enter__() inside a disabled function but does NOT
return a contextmanager object. Therefore it doesn't trigger the
codegen.py code path that was changed from IncorrectUsage to
unimplemented(). It should still expect InternalTorchDynamoError.

The other two tests (test_return_new_contextmanager and
test_return_advanced_contextmanager) correctly expect Unsupported
because they actually return contextmanager objects.

Signed-off-by: Yashwant Bezawada <yashwant_b@me.com>

yashwantbezawada commented:
Fixed the test failure - `test_disable___enter__` doesn't return a contextmanager so it shouldn't have been changed to expect `Unsupported`. Reverted that one test back to `InternalTorchDynamoError`.

@williamwen42 @Lucaskabela mind merging once CI is green?

yashwantbezawada commented:
Closing this. The codebase has moved on since this PR was opened and the contextmanager support has been properly implemented now. The tests expect success rather than exceptions, so this refactor is no longer needed.

Thanks @williamwen42 for the reviews!
